File: scala/ex5_3_2.scala
trait Comparable[T]:
  def gt(v1: T, v2: T): Boolean

given Comparable[Int] with
  def gt(v1: Int, v2: Int): Boolean = v1 > v2

def sort[T: Comparable](lst: List[T]): Unit =
  if summon.gt(lst(0), lst(1)) then ()

given [T: Comparable]: Comparable[List[T]] with
  def gt(v1: List[T], v2: List[T]): Boolean =
    if summon.gt(v1(0), v2(0)) then ()
    true

@main def main =
  sort[List[Int]](List())
  sort[List[List[Int]]](List())

File: rust/ex5_3_2.rs
trait Comparable {
    fn gt(v1: &Self, v2: &Self) -> bool;
}

impl Comparable for i32 {
    fn gt(v1: &i32, v2: &i32) -> bool { v1 > v2 }
}

fn sort<T: Comparable>(lst: Vec<T>) -> () {
    if Comparable::gt(&lst[0], &lst[1]) {}
}

impl<T: Comparable> Comparable for Vec<T> {
    fn gt(v1: &Vec<T>, v2: &Vec<T>) -> bool {
        if Comparable::gt(&v1[0], &v2[0]) {}
        true
    }
}

fn main() {
    sort::<Vec<i32>>(vec![vec![1, 2], vec![1, 2, 3], vec![1]]);
    sort::<Vec<Vec<i32>>>(vec![]);
}
